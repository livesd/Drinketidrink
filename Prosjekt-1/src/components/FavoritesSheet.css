:root {
  --dock-left: 16px;
  --dock-top: 120px;
  --hamburger-size: 48px;
  --dock-gap: 12px;
}

/* hamburger menu */
.hamburger {
  position: fixed;
  left: var(--dock-left);
  top: var(--dock-top);
  z-index: 60;
  width: 48px;
  height: 48px;
  border: 1px solid #d1d5db;
  background: #f7e6ca;
  border-radius: 12px;
  cursor: pointer;
  box-shadow: 0 6px 14px rgba(0, 0, 0, 0.08);
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 5.5px;
}

/* the bars inside the menu button */
.hamburger .bar {
  width: 22px;
  height: 2px;
  background: #1f2937;
  display: block;
}

/* the badge indicating the number of favorites */
.hamburger .badge {
  position: absolute;
  top: -6px;
  right: -6px;
  min-width: 18px;
  height: 18px;
  padding: 0 4px;
  border-radius: 9999px;
  background: #1f2937;
  color: #f7e6ca;
  font-size: 11px;
  line-height: 18px;
  text-align: center;
}

/* background for clicks in area outside the sheet */
.sheet-backdrop {
  pointer-events: none;
}

.sheet-backdrop.open {
  pointer-events: auto;
}

/* favorites list */
.favorites-sheet {
  position: fixed;
  top: calc(var(--dock-top) - 2px);
  left: calc(var(--dock-left) + var(--hamburger-size) + var(--dock-gap));
  right: auto;
  width: min(70vw, 240px);
  max-height: min(60vh, 480px);
  overflow: auto;
  direction: rtl;
  border-radius: 16px;
  background: transparent;
  background: rgb(83, 68, 71);
  transform: translateY(-8px) scale(.98);
  opacity: 0;
  z-index: 50;
  transition: transform 0.3s ease, opacity 0.3s ease;
}

/* changes the direction back to ltr, so the scrollbar is on the left side of the list */
.favorites-sheet>* {
  direction: ltr;
}

/* the symbol in the header of favorites list */
.sheet-header strong::before {
  content: "ðŸ¥‚";
  margin-left: .35rem;
  opacity: .75;
  font-size: 0.9em;
}

.favorites-sheet.open {
  transform: translateY(0) scale(1);
  opacity: 1;
}

/* header of the favorites list */
.sheet-header {
  background: transparent;
  border-bottom: none;
  text-align: left;
  position: sticky;
  top: 0;
  z-index: 1;
  color: #f7e6ca;
  background-color: rgb(83, 68, 71);
  display: inline-block;
  padding-left: 1rem;
}

/* close button of fav list */
.close-btn {
  font-size: 22px;
  line-height: 1;
  cursor: pointer;
  color: #f7e6ca;
}

/* favorites list */
.fav-list {
  list-style: none;
  margin: 0.25rem 0 0;
  padding: 0;
}

/* item in the favorites list, consists of remove button and drink */
.fav-item {
  display: grid;
  grid-template-columns: 28px auto;
  gap: 0.5rem;
  align-items: center;
  padding: 0.25rem 0.25rem;
  border-bottom: transparent;
}

/* the drink of the favorite item */
.fav-item .link {
  justify-self: start;
  width: max-content;
  text-align: left;
  background: transparent;
  font: inherit;
  cursor: pointer;
  padding: 0.25rem 0.25rem;
  color: #f7e6ca;
  grid-column: 2/3;
  grid-row: 1/2;
}

.fav-item .link:hover {
  text-decoration: underline;
}

/* the remove button of the favorite item */
.fav-item .remove {
  width: 28px;
  height: 28px;
  padding: 0;
  display: grid;
  place-items: center;
  font-size: 0;
  grid-column: 1/2;
  grid-row: 1/2;
}

.fav-item .remove::before {
  content: "Ã—";
  font-size: 18px;
  color: #f7e6ca;
  font-weight: bolder;
}

.muted.small {
  color: #374151;
  font-size: .9rem;
}

/* responsive */
@media (max-width: 640px) {
  :root {
    --dock-left: 0;
    --dock-top: 0;
    --dock-gap: 0;
  }

  .favorites-sheet {
    top: 0;
    width: 100vw;
    left: 0;
    max-width: 100vw;
    height: auto;
    max-height: 60vh;
    border-radius: 0;
    transform: translateY(-100%);
  }

  .favorites-sheet.open {
    transform: translateY(0);
  }
}