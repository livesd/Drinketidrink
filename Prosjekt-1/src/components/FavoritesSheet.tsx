// src/components/FavoritesSheet.tsx
import { useEffect, useRef } from "react";

type Fav = { id: string; name: string };

type Props = {
  id?: string;
  open: boolean;
  favorites: Fav[];
  onClose: () => void;
  onSelect: (id: string) => void; // hopp til kort i lista
  onToggleFavorite: (id: string, name: string) => void;
};

export default function FavoritesSheet({
  id = "favorites-sheet",
  open,
  favorites,
  onClose,
  onSelect,
  onToggleFavorite,
}: Props) {
  const ref = useRef<HTMLDivElement>(null);

  // Lukk ved ESC
  useEffect(() => {
    const h = (e: KeyboardEvent) => {
      if (e.key === "Escape") onClose();
    };
    window.addEventListener("keydown", h);
    return () => window.removeEventListener("keydown", h);
  }, [onClose]);

  // Lukk ved klikk utenfor
  useEffect(() => {
    const h = (e: MouseEvent) => {
      if (!open) return;
      if (ref.current && !ref.current.contains(e.target as Node)) onClose();
    };
    document.addEventListener("mousedown", h);
    return () => document.removeEventListener("mousedown", h);
  }, [open, onClose]);

  return (
    <div className={`sheet-backdrop ${open ? "open" : ""}`} aria-hidden={!open}>
      <aside
        id={id}
        ref={ref}
        className={`favorites-sheet ${open ? "open" : ""}`}
        role="dialog"
        aria-modal="false"
        aria-label="Favorites notepad"
      >
        <header className="sheet-header">
          <strong>Favorites</strong>
          <button className="close-btn" onClick={onClose} aria-label="Close">
            Ã—
          </button>
        </header>

        {favorites.length === 0 ? (
          <p className="muted small">No favorites yet.</p>
        ) : (
          <ul className="fav-list" role="list">
            {favorites.map((f) => (
              <li key={f.id} className="fav-item">
                <button
                  className="link"
                  onClick={() => {
                    onSelect(f.id);
                    onClose();
                  }}
                  title={`Vis ${f.name}`}
                >
                  {f.name}
                </button>
                <button
                  className="remove"
                  onClick={() => onToggleFavorite(f.id, f.name)}
                  aria-label={`Remove ${f.name} from favorites`}
                  title="Remove"
                >
                  Remove
                </button>
              </li>
            ))}
          </ul>
        )}
      </aside>
    </div>
  );
}
