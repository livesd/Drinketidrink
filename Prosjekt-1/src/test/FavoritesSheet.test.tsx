import "@testing-library/jest-dom/vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import FavoritesSheet from "../components/FavoritesSheet";

describe("FavoritesSheet", () => {
  const favs = [
    { id: "11000", name: "Mojito" },
    { id: "17222", name: "A1" },
  ];

  it("renders empty state when no favorites", () => {
    const onClose = vi.fn();
    const onSelect = vi.fn();
    const onToggle = vi.fn();

    const { container } = render(
      <FavoritesSheet
        open={true}
        favorites={[]}
        onClose={onClose}
        onSelect={onSelect}
        onToggleFavorite={onToggle}
      />
    );

    expect(screen.getByText(/no favorites yet/i)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("applies 'open' classes and aria attributes when open=true", () => {
    const { container } = render(
      <FavoritesSheet
        open={true}
        favorites={favs}
        onClose={vi.fn()}
        onSelect={vi.fn()}
        onToggleFavorite={vi.fn()}
      />
    );

    const dialog = screen.getByRole("dialog", {
      name: /favorites notepad/i,
      hidden: true,
    });
    expect(dialog).toHaveClass("open");

    const backdrop = container.querySelector(".sheet-backdrop");
    expect(backdrop).toHaveClass("open");
    expect(backdrop).toHaveAttribute("aria-hidden", "false");
  });

  it("does not apply 'open' class when open=false", () => {
    const { container } = render(
      <FavoritesSheet
        open={false}
        favorites={favs}
        onClose={vi.fn()}
        onSelect={vi.fn()}
        onToggleFavorite={vi.fn()}
      />
    );

    // When closed, the sheet is inside aria-hidden backdrop, so include hidden nodes
    const dialog = screen.getByRole("dialog", {
      name: /favorites notepad/i,
      hidden: true,
    });
    expect(dialog).not.toHaveClass("open");

    const backdrop = container.querySelector(".sheet-backdrop");
    expect(backdrop).not.toHaveClass("open");
    expect(backdrop).toHaveAttribute("aria-hidden", "true");
  });

  it("closes on Escape key", () => {
    const onClose = vi.fn();
    render(
      <FavoritesSheet
        open={true}
        favorites={favs}
        onClose={onClose}
        onSelect={vi.fn()}
        onToggleFavorite={vi.fn()}
      />
    );

    fireEvent.keyDown(window, { key: "Escape" });
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it("closes when clicking outside (backdrop/body)", () => {
    const onClose = vi.fn();
    render(
      <FavoritesSheet
        open={true}
        favorites={favs}
        onClose={onClose}
        onSelect={vi.fn()}
        onToggleFavorite={vi.fn()}
      />
    );

    fireEvent.mouseDown(document.body);
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it("does not close when clicking inside the sheet", () => {
    const onClose = vi.fn();
    render(
      <FavoritesSheet
        open={true}
        favorites={favs}
        onClose={onClose}
        onSelect={vi.fn()}
        onToggleFavorite={vi.fn()}
      />
    );

    const dialog = screen.getByRole("dialog", { name: /favorites notepad/i });
    fireEvent.mouseDown(dialog);
    expect(onClose).not.toHaveBeenCalled();
  });

  it("does not close when clicking the hamburger button", () => {
    const onClose = vi.fn();

    const hamburger = document.createElement("button");
    hamburger.className = "hamburger";
    document.body.appendChild(hamburger);

    render(
      <FavoritesSheet
        open={true}
        favorites={favs}
        onClose={onClose}
        onSelect={vi.fn()}
        onToggleFavorite={vi.fn()}
      />
    );

    fireEvent.mouseDown(hamburger);
    expect(onClose).not.toHaveBeenCalled();

    document.body.removeChild(hamburger);
  });

  it("selects a favorite and closes", () => {
    const onClose = vi.fn();
    const onSelect = vi.fn();
    render(
      <FavoritesSheet
        open={true}
        favorites={favs}
        onClose={onClose}
        onSelect={onSelect}
        onToggleFavorite={vi.fn()}
      />
    );

    fireEvent.click(screen.getByRole("button", { name: "Mojito" }));
    expect(onSelect).toHaveBeenCalledWith("11000");
    expect(onClose).toHaveBeenCalled();
  });

  it("removes a favorite via remove button", () => {
    const onToggle = vi.fn();
    render(
      <FavoritesSheet
        open={true}
        favorites={favs}
        onClose={vi.fn()}
        onSelect={vi.fn()}
        onToggleFavorite={onToggle}
      />
    );

    fireEvent.click(
      screen.getByRole("button", { name: /remove mojito from favorites/i })
    );
    expect(onToggle).toHaveBeenCalledWith("11000", "Mojito");
  });

  it("close button triggers onClose", () => {
    const onClose = vi.fn();
    render(
      <FavoritesSheet
        open={true}
        favorites={favs}
        onClose={onClose}
        onSelect={vi.fn()}
        onToggleFavorite={vi.fn()}
      />
    );

    fireEvent.click(screen.getByRole("button", { name: /close/i }));
    expect(onClose).toHaveBeenCalled();
  });

  it("matches snapshot when open with favorites", () => {
    const { container } = render(
      <FavoritesSheet
        open={true}
        favorites={favs}
        onClose={vi.fn()}
        onSelect={vi.fn()}
        onToggleFavorite={vi.fn()}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
