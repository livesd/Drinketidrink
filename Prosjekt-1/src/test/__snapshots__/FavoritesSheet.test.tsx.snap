// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavoritesSheet > matches snapshot when open with favorites 1`] = `
<div
  aria-hidden="false"
  class="sheet-backdrop open"
>
  <aside
    aria-label="Favorites notepad"
    aria-modal="false"
    class="favorites-sheet open"
    id="favorites-sheet"
    role="dialog"
  >
    <header
      class="sheet-header"
    >
      <strong>
        Favorites
      </strong>
      <button
        aria-label="Close"
        class="close-btn"
      >
        ×
      </button>
    </header>
    <ul
      class="fav-list"
      role="list"
    >
      <li
        class="fav-item"
      >
        <button
          class="link"
          title="Vis Mojito"
        >
          Mojito
        </button>
        <button
          aria-label="Remove Mojito from favorites"
          class="remove"
          title="Remove"
        >
          Remove
        </button>
      </li>
      <li
        class="fav-item"
      >
        <button
          class="link"
          title="Vis A1"
        >
          A1
        </button>
        <button
          aria-label="Remove A1 from favorites"
          class="remove"
          title="Remove"
        >
          Remove
        </button>
      </li>
    </ul>
  </aside>
</div>
`;

exports[`FavoritesSheet > renders empty state when no favorites 1`] = `
<div
  aria-hidden="false"
  class="sheet-backdrop open"
>
  <aside
    aria-label="Favorites notepad"
    aria-modal="false"
    class="favorites-sheet open"
    id="favorites-sheet"
    role="dialog"
  >
    <header
      class="sheet-header"
    >
      <strong>
        Favorites
      </strong>
      <button
        aria-label="Close"
        class="close-btn"
      >
        ×
      </button>
    </header>
    <p
      class="muted small"
    >
      No favorites yet.
    </p>
  </aside>
</div>
`;
